rio/api/src/main/java/org/eclipse/rdf4j/rio/helpers/XMLReaderBasedParser.java

```diff
@@ -0,0 +1,169 @@
+import org.eclipse.rdf4j.rio.helpers.XMLParserSettings;
+import org.xml.sax.SAXException;
+import org.xml.sax.SAXNotRecognizedException;
+import org.xml.sax.SAXNotSupportedException;
+import org.xml.sax.XMLReader;
+
+/**
+ * Base class for Rio parsers that are based on a SAX {@link XMLReader}.
+ * 
+ * @author Jeen Broekstra
+ */
+public abstract class XMLReaderBasedParser extends AbstractRDFParser {
+
+	public XMLReaderBasedParser(ValueFactory f) {
+		super(f);
+	}
+
+	/**
+	 * Returns a collection of settings that will always be set as XML parser properties using
+	 * {@link XMLReader#setProperty(String, Object)}
+	 * <p>
+	 * Subclasses can override this to specify more supported settings.
+	 * 
+	 * @return A collection of {@link RioSetting}s that indicate which properties will always be setup using
+	 *         {@link XMLReader#setProperty(String, Object)}.
+	 */
+	public Collection<RioSetting<?>> getCompulsoryXmlPropertySettings() {
+		return Collections.<RioSetting<?>> emptyList();
+	}
+
+	/**
+	 * Returns a collection of settings that will always be set as XML parser features using
+	 * {@link XMLReader#setFeature(String, boolean)}.
+	 * <p>
+	 * Subclasses can override this to specify more supported settings.
+	 * 
+	 * @return A collection of {@link RioSetting}s that indicate which boolean settings will always be setup
+	 *         using {@link XMLReader#setFeature(String, boolean)}.
+	 */
+	public Collection<RioSetting<Boolean>> getCompulsoryXmlFeatureSettings() {
+		Set<RioSetting<Boolean>> results = new HashSet<RioSetting<Boolean>>();
+		results.add(XMLParserSettings.SECURE_PROCESSING);
+		results.add(XMLParserSettings.DISALLOW_DOCTYPE_DECL);
+		results.add(XMLParserSettings.EXTERNAL_GENERAL_ENTITIES);
+		results.add(XMLParserSettings.EXTERNAL_PARAMETER_ENTITIES);
+		return results;
+	}
+
+	/**
+	 * Returns a collection of settings that will be used, if set in {@link #getParserConfig()}, as XML parser
+	 * properties using {@link XMLReader#setProperty(String, Object)}
+	 * <p>
+	 * Subclasses can override this to specify more supported settings.
+	 * 
+	 * @return A collection of {@link RioSetting}s that indicate which properties can be setup using
+	 *         {@link XMLReader#setProperty(String, Object)}.
+	 */
+	public Collection<RioSetting<?>> getOptionalXmlPropertySettings() {
+		return Collections.<RioSetting<?>> emptyList();
+	}
+
+	/**
+	 * Returns a collection of settings that will be used, if set in {@link #getParserConfig()}, as XML parser
+	 * features using {@link XMLReader#setFeature(String, boolean)}.
+	 * <p>
+	 * Subclasses can override this to specify more supported settings.
+	 * 
+	 * @return A collection of {@link RioSetting}s that indicate which boolean settings can be setup using
+	 *         {@link XMLReader#setFeature(String, boolean)}.
+	 */
+	public Collection<RioSetting<Boolean>> getOptionalXmlFeatureSettings() {
+		Set<RioSetting<Boolean>> results = new HashSet<RioSetting<Boolean>>();
```

Comment:
FYI these sets should probably be unmodifiable and maybe declared outside of the method so that we don't instantiate a new set every time. Also, since java 7, you can declare these without the redundat type spec: Set<RioSetting<Boolean>> results = new HashSet<>();

Here is another way of declaring it which I prefer:

```
Set<RioSetting<Boolean>> results = new HashSet<>(
    Arrays.asList(
        XMLParserSettings.LOAD_EXTERNAL_DTD
    )
);
```